proxy-node-gateway/src/test/java/uk/gov/ida/notification/apprule/rules/TestSamlResponseExceptionMapperRule.java

```diff
@@ -0,0 +1,108 @@
+package uk.gov.ida.notification.apprule.rules;
+
+import io.dropwizard.testing.junit.DropwizardClientRule;
+import io.dropwizard.testing.junit.ResourceTestRule;
+import org.glassfish.jersey.client.ClientProperties;
+import org.glassfish.jersey.test.grizzly.GrizzlyWebTestContainerFactory;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.slf4j.MDC;
+import uk.gov.ida.notification.SamlFormViewBuilder;
+import uk.gov.ida.notification.exceptions.mappers.ErrorPageExceptionMapper;
+import uk.gov.ida.notification.exceptions.mappers.ExceptionToSamlErrorResponseMapper;
+import uk.gov.ida.notification.exceptions.mappers.GenericExceptionMapper;
+import uk.gov.ida.notification.proxy.TranslatorProxy;
+import uk.gov.ida.notification.session.GatewaySessionData;
+import uk.gov.ida.notification.session.storage.SessionStore;
+import uk.gov.ida.notification.shared.ProxyNodeMDCKey;
+import uk.gov.ida.notification.views.SamlFormView;
+
+import javax.annotation.RegEx;
+import javax.ws.rs.core.Response;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doCallRealMethod;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class TestSamlResponseExceptionMapperRule {
+
+    private static final SamlFormViewBuilder samlFormViewBuilder = mock(SamlFormViewBuilder.class);
+    private static final TranslatorProxy translatorProxy = mock(TranslatorProxy.class);
+    private static final SessionStore sessionStore = mock(SessionStore.class);
+
+    @Before
+    public void before() {
+        when(sessionStore.getSession(any(String.class))).thenReturn(new GatewaySessionData("HubRequestId","EidasRequestId","EidasDestination","EidasConnectorPublicKey","EidasRelayState"));
+        String nullString = null;
+        when(samlFormViewBuilder.buildResponse("EidasDestination", nullString, "Continue", "EidasRelayState"))
+                .thenReturn(
+                new SamlFormView("postUrl",
+                                 "samlMessageType",
+                                 "encodedSamlMessage",
+                                 "submitText"));
+    }
+
+    @ClassRule
+    public static final ResourceTestRule resources = ResourceTestRule.builder()
+                                    .setTestContainerFactory(new GrizzlyWebTestContainerFactory())
+                                    .addProvider(new ExceptionToSamlErrorResponseMapper(samlFormViewBuilder, translatorProxy, sessionStore))
+                                    .addResource(new TestExceptionMapperResource())
+                                    .build();
+
+    @Test
+    public void shouldMapSessionAlreadyExistsExceptionToExceptionToSamlErrorResponseMapper() {
+
+        var response = resources.getJerseyTest()
```

Comment:
Fancy Java 10 feature! 
I personally like this but we've not used this before so worth bringing this up with the team so we can decide when it's acceptable to use this.

Reply:
I forgot I used this.  I remembered it was a thing and wondered if it would work.  It did and I promptly forgot about it.  I've changed it back for now.  I'm 50/50 on whether I like it or not anyway!

Reply:
I think it's ok in tests but probably not in prod code as it makes it more ambiguous. Feel free to bring this up at a dev meeting or on Slack